<a href="https://colab.research.google.com/github/cw805/MATH-242/blob/main/Prime_Gaps.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Prime gaps

A prime gap is the difference between successive prime numbers. The first few prime gaps are 1, 2, 2, 4, 2, 4, . . .. What is the distribution of prime gaps? Compute the distribution the first n prime
gaps for various choices of n and report on your observations. Do you think there is a most common prime gap? Explain.

In [ ]:
import math   
import sympy
import numpy as np
from collections import Counter 

# As usual, first get a list of primes

We will implement the Sieve of Eratosthenes function for a list of primes as it was the most efficient algorithm for calculating primes we found so far

In [ ]:
def fastEratos(nMax):
  
  nums = list(range(2,nMax+1))
  nroot = math.sqrt(nMax)

  i = 0 
  while nums[i] <= nroot:

    if nums[i] > 0:   
      j = i + nums[i]  
      while j < len(nums):
        nums[j] = 0
        j += nums[i]    

    i += 1 

  primes = [n for n in nums if n != 0]
  return primes

In [ ]:
print(fastEratos(100))
eratoslist = fastEratos(100)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]


### Write a function that generates a list of prime gaps

In [ ]:
def primeGap(nMax):

  gaps = [n for n in fastEratos(nMax)]

  gapList = list(np.diff(gaps)) # np.diff(gaps) is an array

  return gapList

In [ ]:
gapslist = primeGap(1000000)
print(gapslist)

### Count occurences of each gaps and find the most common gaps

I've used these key-value dictionary methods in python before and know that are effective for tasks like these. Here we can use the built-in function Counter() in the collections module. It works by using key-value relationships so that we can count occurences of each "key" which are prime gaps, and then sort by its "value," which are the occurences. Then, we can also sort the keys and see what's the largest key value

In [ ]:
counts = Counter(gapslist)
counts.most_common()

[(6, 13549),
 (2, 8169),
 (4, 8143),
 (12, 8005),
 (10, 7079),
 (8, 5569),
 (18, 4909),
 (14, 4233),
 (16, 2881),
 (24, 2682),
 (20, 2401),
 (22, 2172),
 (30, 1914),
 (28, 1234),
 (26, 1175),
 (36, 767),
 (34, 557),
 (32, 550),
 (42, 476),
 (40, 424),
 (38, 330),
 (44, 202),
 (48, 196),
 (46, 155),
 (54, 140),
 (50, 106),
 (60, 96),
 (52, 77),
 (58, 54),
 (56, 53),
 (66, 48),
 (64, 24),
 (70, 22),
 (62, 16),
 (72, 13),
 (68, 13),
 (78, 13),
 (74, 12),
 (76, 6),
 (84, 6),
 (82, 5),
 (86, 4),
 (90, 4),
 (80, 3),
 (96, 2),
 (100, 2),
 (1, 1),
 (112, 1),
 (114, 1),
 (88, 1),
 (98, 1),
 (92, 1)]

In [ ]:
sorted(counts)

[1,
 2,
 4,
 6,
 8,
 10,
 12,
 14,
 16,
 18,
 20,
 22,
 24,
 26,
 28,
 30,
 32,
 34,
 36,
 38,
 40,
 42,
 44,
 46,
 48,
 50,
 52,
 54,
 56,
 58,
 60,
 62,
 64,
 66,
 68,
 70,
 72,
 74,
 76,
 78,
 80,
 82,
 84,
 86,
 88,
 90,
 92,
 96,
 98,
 100,
 112,
 114]

### Observation


For up to 1 million starting values n, 6 is the most common prime gap, and the largest prime gap is 114.